Misc ray tracing changes

GiudGiud opened the pull request:
misc minor changes I needed for MaCaw

I might add another one based on Michael's input on slack
needs a test

@mjseneca15

refs #26851
refs #26960 

EDIT: this is ready for discussion

GiudGiud committed:
Loosen comparison tolerances for a vertex intersection check that gets hit due to a previous
tolerance
More verbose error

refs #26851

GiudGiud committed:
Avoid overwhelming user with warning and output one problematic element

GiudGiud commented:
looks like some of these changes are not good for existing tests. I ll take this on later

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2069448/) on 54b150c wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/26852/site)

This comment will be updated on new commits.

GiudGiud committed:
Do not consider the incoming element to be a neighbor
this only triggers checking exitElems twice, which fails twice in that situation

GiudGiud committed:
Make sure sampled directions do not leave protruding faces
by not running those rays

Apply suggestions from code review

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

moosebuild commented:
Job [Modules debug](https://civet.inl.gov/job/2094016/) on 91e5f47 : invalidated by @GiudGiud

fluke?

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2094042/) on 91e5f47 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>550348</th>
      <th colspan=4>#26852 91e5f4</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.29%</td>
      <td>85.29%</td>
      <td>+0.00%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>100890</td>
      <td>100892</td>
      <td>+2</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>17401</td>
      <td>17399</td>
      <td>-2</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/26852/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/26852/coverage/framework)


## Modules coverage

### Fluid properties

<table>
  <thead>
    <tr>
      <th></th>
      <th>550348</th>
      <th colspan=4>#26852 91e5f4</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.47%</td>
      <td>85.46%</td>
      <td>-0.01%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>7258</td>
      <td>7257</td>
      <td>-1</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1234</td>
      <td>1235</td>
      <td>+1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/26852/coverage_diff/fluid_properties)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/26852/coverage/fluid_properties)

### Heat transfer

<table>
  <thead>
    <tr>
      <th></th>
      <th>550348</th>
      <th colspan=4>#26852 91e5f4</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>88.46%</td>
      <td>88.05%</td>
      <td>-0.41%</td>
      <td>18.52%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>4055</td>
      <td>4060</td>
      <td>+5</td>
      <td>5</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>529</td>
      <td>551</td>
      <td>+22</td>
      <td>22</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/26852/coverage_diff/heat_transfer)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/26852/coverage/heat_transfer)

### Ray tracing

<table>
  <thead>
    <tr>
      <th></th>
      <th>550348</th>
      <th colspan=4>#26852 91e5f4</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>95.29%</td>
      <td>95.29%</td>
      <td>-0.00%</td>
      <td>90.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>4326</td>
      <td>4325</td>
      <td>-1</td>
      <td>9</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>214</td>
      <td>214</td>
      <td>-</td>
      <td>1</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/26852/coverage_diff/ray_tracing)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/26852/coverage/ray_tracing)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/26852/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/26852/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/26852/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/26852/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/26852/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/26852/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/26852/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/26852/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/26852/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/26852/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/26852/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/26852/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/26852/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/26852/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/26852/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/26852/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/26852/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/26852/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/26852/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/26852/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/26852/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/26852/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/26852/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/26852/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/26852/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/26852/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/26852/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/26852/coverage/xfem)

</details>

## Warnings

- `heat_transfer` new line coverage rate 18.52% is less than the suggested 90.0%



This comment will be updated on new commits.

lindsayad commented:
seems like not a fluke

GiudGiud commented:
Yeah it s gotten worse

GiudGiud commented:
Should be fixed now.

lindsayad commented on modules/heat_transfer/src/userobjects/ViewFactorRayStudy.C:
```diff
@@ -312,6 +314,68 @@ ViewFactorRayStudy::generateRays()
+          }
+
+          // Correct the direction if needed
+          auto grazing_dir = (intersection_point - start_elem._points[start_i]).unit();
```
```suggestion
          const auto grazing_dir = (intersection_point - start_elem._points[start_i]).unit();
```

lindsayad commented on modules/heat_transfer/src/userobjects/ViewFactorRayStudy.C:
```diff
@@ -312,6 +314,68 @@ ViewFactorRayStudy::generateRays()
+          auto grazing_dir = (intersection_point - start_elem._points[start_i]).unit();
+          if (intersection_found && inward_normal * direction < inward_normal * grazing_dir)
+          {
+            const Real nd = inward_normal * direction;
+            const Real gn = inward_normal * grazing_dir;
+            const Real b = sqrt((1 - gn * gn) / (1 - nd * nd));
```
some more informative variables names or comments would be helpful for the layman

lindsayad commented on modules/ray_tracing/include/raytracing/TraceRayTools.h:
```diff
@@ -186,16 +188,15 @@ findNeighbors(
-      for (auto current_neighbor : elem->neighbor_ptr_range())
-        if (current_neighbor &&
-            current_neighbor != remote_elem) // we have a real neighbor on elem side
+    for (const Elem * test_elem : untested_set)
```
```suggestion
    for (const Elem * const test_elem : untested_set)
```

lindsayad commented on modules/ray_tracing/include/raytracing/TraceRayTools.h:
```diff
@@ -186,16 +188,15 @@ findNeighbors(
-        if (current_neighbor &&
-            current_neighbor != remote_elem) // we have a real neighbor on elem side
+    for (const Elem * test_elem : untested_set)
+      for (auto current_neighbor : test_elem->neighbor_ptr_range())
```
```suggestion
      for (const auto * const current_neighbor : test_elem->neighbor_ptr_range())
```

lindsayad commented on modules/ray_tracing/include/raytracing/TraceRayTools.h:
```diff
@@ -594,9 +596,11 @@ sideIntersectedByLine(const Elem * elem,
-                "Doesn't intersect vertex");
+    mooseAssert(
+        intersected_extrema.vertexPoint(elem).absolute_fuzzy_equals(intersection_point,
+                                                                    3 * TRACE_TOLERANCE),
```
magic number 😄

lindsayad reviewed (commented):
this intersection stuff doesn't appear to consider curved surfaces (quadratic elements) but that's fine. It would likely introduce a lot of complexity to do so. Just noting for myself largely

GiudGiud committed:
Add a test with a nonplanar face that has rays exiting

loganharbour commented on modules/heat_transfer/src/userobjects/ViewFactorRayStudy.C:
```diff
@@ -312,6 +314,68 @@ ViewFactorRayStudy::generateRays()
+        // non-planar side.
+        // TODO: modify the weights to keep the quadrature order
+        bool intersection_found = false;
+        if (_is_3d && start_elem._start_elem &&
```
We should only call `side_ptr` once; it builds a whole new element. there's a helper that does this to avoid construction

GiudGiud commented:
Now that I have a test I m reworking this. I ll fix the side ptr calls

GiudGiud commented:
The more I think about it, the more changing the direction is not the right thing to do. 
It will bias results. Killing the ray or never starting it since it points the wrong direction will give better results

GiudGiud commented:
fixed. much better now

GiudGiud commented:
nevermind, getting a lot of failures I m not seeing locally

GiudGiud commented:
good now, maybe

GiudGiud commented:
Nice. Test failure resolved
